[NFC][PowerPC] Add test to check lanemasks for subregisters.

stefanp-synopsys opened the pull request:
This change adds a test case to check the lane masks for a varitey of
subregisters.

stefanp-synopsys committed:
[NFC][PowerPC] Add test to check lanemasks for subregisters.

arsenm commented on llvm/test/CodeGen/PowerPC/subreg-lanemasks.mir:
```diff
@@ -0,0 +1,67 @@
+# RUN: llc -mcpu=pwr10 -O3 -ppc-track-subreg-liveness -verify-machineinstrs \
+# RUN:   -mtriple=powerpc64le-unknown-linux-gnu -run-pass=greedy,virtregrewriter \
+# RUN:   -debug-only=regalloc -o - %s 2>&1 >/dev/null | FileCheck %s
```
Use -filetype=null instead of outputting to /dev/null. Also don't need -O3 or -verify-machineinstrs. 

Also needs REQUIRES: asserts if you're checking the debug output

stefanp-ibm replied:
Thank you for looking at this patch.
I've made the changes that you mentioned.

stefanp-synopsys committed:
[PowerPC] Updated run lines in the test.

Removed a couple of options from the run line that were not required.
Added the REQUIRES: asserts.

arsenm commented on llvm/test/CodeGen/PowerPC/subreg-lanemasks.mir:
```diff
@@ -0,0 +1,68 @@
+
+# Keep track of all of the lanemasks for various subregsiters.
+#
+# TODO: The mask for %6.sub_vsx1:accrc is the same as the mask for %10.sub_vsx1_then_sub_64:accrc
+#       even though one is a 128 bit register and the other is a 64 bit subregister. This should
```
Just because the bitwidths are different doesn't mean this is wrong. They would have the same lane mask if one was half of the other

stefanp-ibm replied:
I'm going to change the comment before I commit. I will shorten it to say that on PPC we want the mask for `%6.sub_vsx1:accrc` and `%10.sub_vsx1_then_sub_64:accrc` to be different.

However, I'm confused about how this is supposed to work. On PPC we have something like this:
```
| ------------- VSRL 128 Bits ------------ |
| --- FPR 64 Bit --- | - Not Used 64 Bit - |
```

The VSRL is marked as `sub_vsx1:accrc` while the FPR is marked as `sub_vsx1_then_sub_64:accrc`. 
At one point we end up needing a COPY for a live range split and based on the lane masks we end up copying only the first 64 bits which leads to incorrect results. The solution on our end is to add a phony register for the section that is not used and this then creates different lane masks for the two subregisters. Is this how we are supposed to approach this issue?  Is it not an issue in general when we generate COPY instrs that two subregsiters have different sizes?

arsenm replied:
I've never really understood the generated subregister index logic. I'm not sure what "VSRL is marked as sub_vsx1:accrc" means

stefanp-synopsys committed:
Minor change to a comment.

fmayer commented:
Also broke sanitizer bots: https://lab.llvm.org/buildbot/#/builders/169/builds/40
